node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/ConfigServerApiImpl.java

```diff
@@ -48,15 +54,72 @@
      */
     private volatile SelfCloseableHttpClient client;
 
-    public ConfigServerApiImpl(Collection<URI> configServerUris) {
-        this(configServerUris, SSLConnectionSocketFactory.getSocketFactory());
+    /**
+     * Creates an api for talking to the config servers.
+     *
+     * <p>Registers with a SslConnectionSocketFactoryUpdater to keep the socket factory
+     * up to date as e.g. any client certificate expires (and unregisters at {@link #close()})
+     */
+    public static ConfigServerApiImpl create(ConfigServerInfo configServerInfo,
+                                             SslConnectionSocketFactoryUpdater updater) {
+        return createFor(updater, configServerInfo.getConfigServerUris());
+    }
+
+    /**
+     * Creates an api for talking to a single config server.
+     *
+     * <p>Registers with a SslConnectionSocketFactoryUpdater to keep the socket factory
+     * up to date as e.g. any client certificate expires (and unregisters at {@link #close()})
+     */
+    public static ConfigServerApiImpl createFor(ConfigServerInfo configServerInfo,
+                                                SslConnectionSocketFactoryUpdater updater,
+                                                HostName configServer) {
+        URI uri = configServerInfo.getConfigServerUri(configServer.value());
+        return createFor(updater, Collections.singletonList(uri));
+    }
+
+    /**
+     * Creates an api for talking to the config servers with a fixed socket factory.
+     *
+     * <p>This should only be necessary for certificate signing requests (CSR) against
+     * the config server when client validation is enabled in the config server.
+     */
+    public static ConfigServerApiImpl createWithSocketFactory(
+            ConfigServerInfo configServerInfo,
+            SSLConnectionSocketFactory socketFactory) {
+        return new ConfigServerApiImpl(configServerInfo.getConfigServerUris(), socketFactory);
+    }
+
+    public static ConfigServerApiImpl createForTestingRemote(URI configServerUri) {
+        return new ConfigServerApiImpl(
+                Collections.singletonList(configServerUri),
+                SSLConnectionSocketFactory.getSocketFactory());
+    }
+
+    public static ConfigServerApiImpl createForTestingWithClient(List<URI> configServerHosts,
```

Comment:
Not a fan of creating public static methods for testing.. At least one of them can be removed by making the constructor package private, the other can be replaced with `createWithSocketFactory()`

Reply:
> Not a fan of creating public static methods for testing..

Why?

Used static here to use the same as other ways of constructing it. 

If someone were to try to find a constructor to use by completing "new ConfigServerApiImpl(", they will get no valid (and misleading) suggestions.

> At least one of them can be removed by making the constructor package private, 
> the other can be replaced with createWithSocketFactory()

Fixed

Reply:
> Why?
> Used static here to use the same as other ways of constructing it.

Yeah, I guess its the `public` I have most issue with, tests can afford to do the setup in a slightly more complicated way rather than have some convenience method that otherwise pollutes the interface.

> If someone were to try to find a constructor to use by completing "new ConfigServerApiImpl(", they will get no valid (and misleading) suggestions.

The remaining method is used by the `ConfigServerApiTest`, so it can be made package private and wont be seen much?

Reply:
With my version of IntelliJ, it completes to "new ConfigServerApiImpl()" w/error with only private constructors, but to "new ConfigServerApiImpl(" with a parameter list that includes all constructors, even private, if at least one them is non-private. This is trying to complete outside of the package.